front/temporal/agent_loop/lib/utils.ts

```diff
@@ -0,0 +1,199 @@
+    step,
+    agentConfiguration,
+    agentMessage,
+    model,
+    prompt,
+    llm,
+    runConfig,
+    publishAgentError,
+  }: GetOutputRequestParams & { llm: LLM | null }
+) {
+  if (llm === null) {
+    return getOutputFromAction(auth, {
+      modelConversationRes,
+      conversation,
+      userMessage,
+      runConfig,
+      specifications,
+      flushParserTokens,
+      contentParser,
+      agentMessageRow,
+      step,
+      agentConfiguration,
+      agentMessage,
+      model,
+      publishAgentError,
+      prompt,
+    });
+  } else {
+    if (userMessage.rank === 0) {
+      // Log conversations that are using the new LLM router (log only once when the conversation starts)
+      localLogger.info(
+        {
+          conversationId: conversation.sId,
+        },
+        "Running model with the new LLM router"
+      );
+    }
+    return getOutputFromLLMStream(auth, {
+      modelConversationRes,
+      conversation,
+      userMessage,
+      runConfig,
+      specifications,
+      flushParserTokens,
+      contentParser,
+      agentMessageRow,
+      step,
+      agentConfiguration,
+      agentMessage,
+      model,
+      publishAgentError,
+      prompt,
+      llm,
+    });
+  }
+}
+
+export async function getOutputFromLLMWithParallelComparisonMode(
+  auth: Authenticator,
+  localLogger: Logger,
+  {
+    modelConversationRes,
+    conversation,
+    userMessage,
+    specifications,
+    flushParserTokens,
+    contentParser,
+    agentMessageRow,
+    step,
+    agentConfiguration,
+    agentMessage,
+    model,
+    prompt,
+    llm,
+    runConfig,
+    publishAgentError,
+  }: GetOutputRequestParams & { llm: LLM }
+) {
+  // Comparison mode: run both implementations in parallel
+  localLogger.info(
+    {
+      conversationId: conversation.sId,
+      step,
+    },
+    "Running LLM comparison mode: executing both old and new implementations in parallel"
+  );
+
+  // We need to create a separate content parser for the new implementation
+  // to avoid conflicts with the old implementation's parser
+  const newContentParser = new AgentMessageContentParser(
+    agentConfiguration,
+    agentMessage.sId,
+    getDelimitersConfiguration({ agentConfiguration })
+  );
+
+  // Helper function to flush tokens for the new parser (we won't publish these events)
+  async function flushNewParserTokens(): Promise<void> {
+    // Consume tokens but don't publish them
+    for await (const _tokenEvent of newContentParser.flushTokens()) {
+      // Silently discard
+    }
+  }
+
+  const [oldResponse, newResponse] = await Promise.all([
+    getOutputFromAction(auth, {
+      modelConversationRes,
+      conversation,
+      userMessage,
+      runConfig,
+      specifications,
+      flushParserTokens,
+      contentParser,
+      agentMessageRow,
+      step,
+      agentConfiguration,
+      agentMessage,
+      model,
+      publishAgentError,
+      prompt,
+    }),
+    getOutputFromLLMStream(auth, {
```

Comment:
I think with this there will be conflicts in the ui no ?

Reply:
Weird, I don't have this in my tests 🤔 

Looking